libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,453 @@
+      [&](const BarDivision& barDivision) {
+         return GetTimeSignature(barDivision, numTatums) ==
+                TimeSignature::FourFour;
+      });
+   // If there is a possibility that this is a 4/4, we don't take any risk and
+   // use this. When we get more clever and use the beat contents, we may be
+   // more assertive.
+   const auto winnerIndex = fourFourIt == possibleBarDivisions.end() ?
+                               GetBestBarDivisionIndex(
+                                  possibleBarDivisions, audioFileDuration,
+                                  numTatums, odf, debugOutput) :
+                               fourFourIt - possibleBarDivisions.begin();
+   const auto& barDivision = possibleBarDivisions[winnerIndex];
+   const auto numBeats = barDivision.numBars * barDivision.beatsPerBar;
+   const auto signature = GetTimeSignature(barDivision, numTatums);
+   const auto bpm = 60. * numBeats / audioFileDuration;
+
+   return { bpm, signature };
+}
+
+bool IsSingleEvent(
+   const std::vector<int>& peakIndices, const std::vector<float>& peakValues)
+{
+   // Detect single-event recordings, e.g. only just one crash cymbal hit. This
+   // will translate as one large peak and maybe a few much smaller ones. In
+   // that case we can expect the average to be between these two groups.
+   const auto peakAvg =
+      std::accumulate(peakValues.begin(), peakValues.end(), 0.) /
+      peakIndices.size();
+   const auto numPeaksAboveAvg =
+      std::count_if(peakValues.begin(), peakValues.end(), [&](float v) {
+         return v > peakAvg;
+      });
+   return numPeaksAboveAvg <= 1;
+}
+} // namespace
+
+std::optional<MusicalMeter> GetMeterUsingTatumQuantizationFit(
+   const MirAudioReader& audio, FalsePositiveTolerance tolerance,
+   const std::function<void(double)>& progressCallback,
+   QuantizationFitDebugOutput* debugOutput)
+{
+   const auto odf =
+      GetOnsetDetectionFunction(audio, progressCallback, debugOutput);
+   const auto odfSr =
+      1. * audio.GetSampleRate() * odf.size() / audio.GetNumSamples();
+   const auto audioFileDuration =
+      1. * audio.GetNumSamples() / audio.GetSampleRate();
+
+   const auto peakIndices = GetPeakIndices(odf);
+   if (debugOutput)
+   {
+      debugOutput->audioFileDuration = audioFileDuration;
+      debugOutput->odfSr = odfSr;
+      debugOutput->odfPeakIndices = peakIndices;
+   }
+
+   const auto peakValues = ([&]() {
+      std::vector<float> peakValues(peakIndices.size());
+      std::transform(
+         peakIndices.begin(), peakIndices.end(), peakValues.begin(),
+         [&](int i) { return odf[i]; });
+      return peakValues;
+   })();
+
+   if (IsSingleEvent(peakIndices, peakValues))
+      return {};
+
+   const auto possibleDivs = GetPossibleDivHierarchies(audioFileDuration);
+   if (possibleDivs.empty())
+      // The file is probably too short to be a loop.
+      return {};
+
+   const auto possibleNumTatums = [&]() {
```

Comment:
This is used only in `RunQuantizationExperiment` -- maybe this computation could be lowered into that function for clarity; pass `possibleDivs` into it instead